stan/math/rev/arr/functor/coupled_ode_system.hpp

```diff
@@ -499,19 +485,28 @@ struct coupled_ode_system<F, var, var> {
 
       for (size_t i = 0; i < N_; i++) {
         dz_dt[i] = dy_dt_vars[i].val();
-        set_zero_all_adjoints_nested();
-        dy_dt_vars[i].grad(z_vars, grad);
+        dy_dt_vars[i].grad();
 
-        for (size_t j = 0; j < N_ + M_; j++) {
+        for (size_t j = 0; j < N_; j++) {
           // orders derivatives by equation (i.e. if there are 2 eqns
           // (y1, y2) and 2 parameters (a, b), dy_dt will be ordered as:
           // dy1_dt, dy2_dt, dy1_da, dy2_da, dy1_db, dy2_db
-          double temp_deriv = j < N_ ? 0 : grad[j];
+          double temp_deriv = 0;
           for (size_t k = 0; k < N_; k++)
-            temp_deriv += z[N_ + N_ * j + k] * grad[k];
+            temp_deriv += z[N_ + N_ * j + k] * y_vars[k].adj();
 
           dz_dt[N_ + i + j * N_] = temp_deriv;
         }
+
+        for (size_t j = 0; j < M_; j++) {
+          double temp_deriv = theta_vars[j].adj();
+          for (size_t k = 0; k < N_; k++)
+            temp_deriv += z[N_ + N_ * N_ + N_ * j + k] * y_vars[k].adj();
+
+          dz_dt[N_ + N_ * N_ + i + j * N_] = temp_deriv;
```

Comment:
Why is this `+ i` instead of `+ k`?  It keeps resetting the same value in the loop as written.